lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,249 @@
+    // If the DeviceManager failed to execute the node, call the callback
+    // provided when ThreadPoolExecutor::run() was called, clean up the
+    // lock and execution state for the run and exit.
+    DoneCb cb = executionState->cb;
+    cb(runId, resultCode, nullptr);
+    executionStates_.erase(runId);
+    executionStateLocks_.erase(runId);
+  } else {
+    // The DeviceManager did not fail to execute the node. This lock is no
+    // longer needed since executionStateLocks_ won't be modified.
+    locksMtxLock.unlock();
+
+    // Erase the input Context for the node because it is no longer needed.
+    (executionState->inputCtxs).erase(node);
+
+    if ((node->children).empty()) {
+      // If the node has no children, propagate its outputs to the result
+      // Context for the run.
+      propagateOutputPlaceholders(runId, std::move(ctx));
+    } else {
+      // If the node has children, propagate its outputs to the input Contexts
+      // for any of its children that need them as inputs.
+      for (auto &child : node->children) {
+        propagatePlaceholdersForNode(runId, child, ctx.get());
+
+        // Execute any children that has no parent nodes left to execute.
+        (executionState->nodeParentsDone)[child] += 1;
+        if ((executionState->nodeParentsDone)[child] ==
+            (child->parents).size()) {
+          executeDAGNode(runId, child);
+        }
+      }
+    }
+
+    // Clean up all the state associated with the node that finished
+    // executing.
+    (executionState->inflightNodes).erase(node);
+    (executionState->nodeParentsDone).erase(node);
+    (executionState->inputCtxs).erase(node);
+
+    // Now, check if all nodes in the graph are done. If so, the callback can be
+    // called and all state associated with the run can be erased.
+
+    // The code at the top of this function acquires locksMtxLock, and then
+    // executionStateLock, so they must be acquired in the same order to avoid
+    // deadlock. Unlock executionStateLock and acquire the locksMtxLock.
+    executionStateLock.unlock();
+    locksMtxLock.lock();
+
+    // Make sure the state for the run was not cleaned up while the
+    // executionStateLock was unlocked.
+    executionStateLockIt = executionStateLocks_.find(runId);
+    if (executionStateLockIt == executionStateLocks_.end()) {
+      // This means another thread snuck in while executionStateLock was
+      // briefly unlocked and cleaned up the state for the run because
+      // another node in the graph failed to execute. Do nothing.
+      return;
+    }
+
+    // Lock executionStateLock before accessing the shared execution state.
+    executionStateLock.lock();
+
+    if ((executionState->inflightNodes).empty()) {
+      // If there are no nodes inflight, that means all nodes are done. Call
+      // the callback and erase the state information.
+      executionState->cb(runId, ResultCode::EXECUTED,
```

Comment:
Holding a lock while calling the callback is dangerous, this could deadlock if the cb calls back into the Executor.